src/Servers/Kestrel/Core/src/Internal/Infrastructure/TransportManager.cs

```diff
@@ -66,6 +67,22 @@ public async Task<EndPoint> BindAsync(EndPoint endPoint, MultiplexedConnectionDe
                     ApplicationProtocols = new List<SslApplicationProtocol>() { new SslApplicationProtocol("h3"), new SslApplicationProtocol("h3-29") }
                 };
 
+                if (listenOptions.HttpsOptions.ServerCertificateSelector != null)
+                {
+                    // We can't set both
+                    sslServerAuthenticationOptions.ServerCertificate = null;
+                    sslServerAuthenticationOptions.ServerCertificateSelectionCallback = (sender, host) =>
+                    {
+                        // There is no ConnectionContext available durring the QUIC handshake.
+                        var cert = listenOptions.HttpsOptions.ServerCertificateSelector(null, host);
```

Comment:
Is there anyway to associate an invocation of this callback with a specific QuicConnection instance? Would it make sense to create a fake ConnectionContext and copy the fake's FeatureCollection to the QuicConnectionContext? The Transport property and the Abort methods on the fake would likely need to throw, but I imagine the most common thing to use the ConnectionContext for in this callback is the FeatureCollection.

It seems better to try to at least partially fake it rather than make existing code referencing the ConnectionContext in the callback nullref if possible.

Reply:
Unfortunately not. The callback is invoked before the QuicConnection is even created and returned from Accept.
https://github.com/dotnet/aspnetcore/blob/8ecc77a1184c6e783595ede4b754af893da496c4/src/Servers/Kestrel/Transport.Quic/src/Internal/QuicConnectionListener.cs#L56-L57

If we made a fake context it would be completely empty.

I think we have to design some new callbacks in .NET 7 at both the System.Net.Quic and Kestrel levels that provide some kind of usable state in these callbacks.

Reply:
> If we made a fake context it would be completely empty.

That would be better than nothing if we copy features that were added out the the later-created QuicConnectionContext. But if we cannot even associate it with the later-created QuicConnectionContext, I guess we're stuck. Maybe we should to update the QuicListener API to make this possible somehow.

Reply:
I don't think even associating the features is possible. I only added the callback in it's current state because I think it's still better than not supporting SNI at all, most use cases only require the host name.

> Maybe we should to update the QuicListener API to make this possible somehow.

Exactly, that's the .NET 7 task.